# Day 4: Passport Processing

#### Zach Bogart

- https://adventofcode.com/2020/day/4

## Setup

- had to pull in the data, split it up, and make it into a list of dictionaries

In [88]:
import re
import pandas as pd
pd.set_option("display.max_rows", 250)

### Read in File

In [2]:
with open("inputs/04-input.txt", "r") as f:
    raw = f.readlines()

In [3]:
# make one big string
one_string = "".join(raw)

In [4]:
# split up with different delimiters
one_string = re.sub("\n\n", "---", one_string)
one_string = re.sub("\n$", "", one_string)
one_string = re.sub("\n", ",", one_string)
one_string = re.sub(" ", ",", one_string)

In [92]:
one_string[:500]

'byr:1983,iyr:2017,pid:796082981,cid:129,eyr:2030,ecl:oth,hgt:182cm---iyr:2019,cid:314,eyr:2039,hcl:#cfa07d,hgt:171cm,ecl:#0180ce,byr:2006,pid:8204115568---byr:1991,eyr:2022,hcl:#341e13,iyr:2016,pid:729933757,hgt:167cm,ecl:gry---hcl:231d64,cid:124,ecl:gmt,eyr:2039,hgt:189in,pid:#9c3ea1---ecl:#1f58f9,pid:#758e59,iyr:2022,hcl:z,byr:2016,hgt:68,eyr:1933---hcl:#fffffd,ecl:gry,eyr:2022,hgt:172cm,pid:781914826,byr:1930,iyr:2018---hcl:#08df7e,ecl:grn,byr:1942,eyr:2028,iyr:2011,cid:141,pid:319110455,hgt:'

In [6]:
# make list of strings describing each dict
dict_strings = one_string.split("---")

In [7]:
# unpack each string into a dict, append
passports = []
for string in dict_strings:
    result = dict((key, val)
          for key, val in (item.split(':')
          for item in string.split(",")))
    passports.append(result)

In [8]:
passports[:3]

[{'byr': '1983',
  'iyr': '2017',
  'pid': '796082981',
  'cid': '129',
  'eyr': '2030',
  'ecl': 'oth',
  'hgt': '182cm'},
 {'iyr': '2019',
  'cid': '314',
  'eyr': '2039',
  'hcl': '#cfa07d',
  'hgt': '171cm',
  'ecl': '#0180ce',
  'byr': '2006',
  'pid': '8204115568'},
 {'byr': '1991',
  'eyr': '2022',
  'hcl': '#341e13',
  'iyr': '2016',
  'pid': '729933757',
  'hgt': '167cm',
  'ecl': 'gry'}]

# Part 1

- check for 8 keys OR 7 keys but missing `cid`

In [59]:
valid_passports_1 = []
for entry in passports:
    num_fields = len(entry.keys())
#     print(num_fields)
    if ((num_fields == 8) or 
        (num_fields == 7 and 'cid' not in entry.keys())):
            valid_passports_1.append(entry)

In [61]:
len(valid_passports_1)

239

# Part 2

- a lot more to check!
- **got tripped up with regular expressions**: Had one more than correct answer (189). Happened to input 188 to see a new error message, but ended up getting the right answer! Took a while to track down looking at the results as a pandas dataframe. Had to me more explicit with the regex for `pid` in particular (could have happened for any of them initially). Didn't specify bounds of string with `^$` so had a `pid` with ten digits that went unnoticed. Dang! Super sneaky bug!
    - Be explicit with regex!

In [83]:
valid_passports_2 = []
for entry in valid_passports_1:
    num_fields = len(entry.keys())
    if ((num_fields == 8) or 
        (num_fields == 7 and 'cid' not in entry.keys())):
            
            # byr (birth year)
            byr = int(entry['byr'])
            if byr < 1920 or byr > 2002:
                continue
                
            # iyr (issue year)
            iyr = int(entry['iyr'])
            if iyr < 2010 or iyr > 2020:
                continue
            
            # eyr (expiration year)
            eyr = int(entry['eyr'])
            if eyr < 2020 or eyr > 2030:
                continue
            
            # hgt (height)
            hgt = entry['hgt']
            if not re.match("^\d+[a-z]{2}$", hgt):
                continue
            
            hgt_unit = entry['hgt'][-2:]
            hgt_value = int(entry['hgt'][:-2])
            
            if hgt_unit not in ['cm', 'in']:
                continue
            
            if hgt_unit == "cm":
                if hgt_value < 150 or hgt_value > 193:
                    continue
            if hgt_unit == "in":
                if hgt_value < 59 or hgt_value > 76:
                    continue
            
            # hcl (hair color)
            hcl = entry['hcl']
            if not (re.match("^#[0-9a-f]{6}$", hcl)):
                continue
        
            # ecl (eye color)
            ecl = entry['ecl']
            if ecl not in ['amb','blu','brn','gry','grn','hzl','oth']:
                continue
                
            # pid (passport id)
            pid = entry['pid']
            if not (re.match("^[0-9]{9}$", pid)):
                continue
                
            # at the end, it is valid    
            valid_passports_2.append(entry)
            

In [85]:
len(valid_passports_2)

188

# Results

### Part 1

In [89]:
pd.DataFrame(valid_passports_1)

iyr  cid   eyr      hcl    hgt      ecl   byr         pid
0    2019  314  2039  #cfa07d  171cm  #0180ce  2006  8204115568
1    2016  NaN  2022  #341e13  167cm      gry  1991   729933757
2    2022  NaN  1933        z     68  #1f58f9  2016     #758e59
3    2018  NaN  2022  #fffffd  172cm      gry  1930   781914826
4    2011  141  2028  #08df7e  186cm      grn  1942   319110455
5    2005  163  1994   2964fb  170cm  #a08502  2005       188cm
6    2013  NaN  2030  #fffffd  168cm      grn  1928   705547886
7    2019  219  2029  #602927  163cm      oth  1943   016251942
8    2010  NaN  2026  #efcc98  184cm      gry  1942   117647952
9    2011  243  2027  #888785  154cm      blu  1962   362697676
10   2016  NaN  2030  #435634  154cm      blu  1965   779104554
11   1993  NaN  1996        z   62cm      zzz  2017     #50f768
12   2016  NaN  2030  #cfa07d  167cm      grn  1988   951967790
13   2018  281  2033  #06a58b   76cm      zzz  2004   320348494
14   2016   83  2028  #341e13  179cm      brn  1941   806979833
15   2018  NaN  2025  #866857   70in      oth  2000   203320330
16   2014  228  2035  #0ec4fe   70cm  #c8533a  2015     #218eb5
17   2017  330  2020  #6b5442  173cm      hzl  1988   173148327
18   2011  NaN  2028  #c0946f     83      grn  1964   931162400
19   2015  163  2027  #efcc98  154cm      oth  1972   838964596
20   2020  NaN  2028  #623a2f  181cm      blu  1976   853714682
21   2013  225  2025  #a97842  173cm      blu  1957   511588647
22   2016  220  2028  #efcc98  177cm      oth  1993   871652492
23   2018   57  2028  #733820  165cm      hzl  1973   018982018
24   2012  NaN  2040  #b6652a  175cm  #c85046  1981   491710153
25   2011  NaN  2028  #9d2ec4   63in      blu  1960   149288934
26   2014  221  2024  #7d3b0c  193cm      oth  1923   705818464
27   2011  NaN  2030  #ceb3a1  157cm      blu  1967   117111015
28   1976  277  2029  #341e13  159in  #6c7644  1965       157cm
29   2018  NaN  2020  #f5dfb9  188cm      brn  1980   787186482
30   2013  168  2023  #07c809  169cm      gry  1945   250679100
31   2013  NaN  2023  #6b5442  171cm      hzl  1965   683134187
32   2016  314  2028  #602927  180cm      blu  1952   720794393
33   2016  NaN  2027  #ddbf30  156cm      brn  1982   185583837
34   2013  NaN  2026  #ceb3a1  175cm      grn  1988   987624973
35   2015  NaN  2028  #a97842  160cm      oth  1974   350988773
36   2013  NaN  2028  #b6652a  156cm      gry  1970   717504683
37   2015  132  2025  #a97842  175cm      hzl  1986   453874703
38   2015  NaN  2026  #7d3b0c  154cm      brn  1959   580659686
39   2018  NaN  2023  #c0946f  191cm      amb  1994   681417707
40   2018   89  2022  #a97842  184cm      blu  1987   302326561
41   2011  NaN  2020  #866857  159cm      blu  1959   457081226
42   2012  NaN  2026  #c0946f  176cm      blu  1964   524745721
43   2014   88  2027  #efcc98  189cm      blu  1974   679155617
44   2020  NaN  2029  #733820  190cm      blu  1935   509009432
45   2019  146  2024  #fffffd  159cm      brn  1931   446558583
46   2014  NaN  2028  #efcc98   65cm      oth  1972   330588516
47   2017  197  2023  #0441c6  175cm      blu  1999   068138358
48   2018  NaN  2022  #866857  193cm      brn  1992   905992465
49   2010  NaN  2030  #c0946f  161cm      gry  1985   616639221
50   2011  260  2023  #a97842  172cm      hzl  1937   594747312
51   2015   83  2028  #fffffd  157cm      brn  1928   134536806
52   2016  195  2027  #583d02  180cm      amb  1934   035433923
53   2010  130  1936   50e1a7   74cm      gmt  1969    82008848
54   2018  NaN  2020  #733820  174cm      hzl  1983   087167304
55   2013  NaN  2020  #888785  188cm      hzl  1972   593301831
56   2020  282  2029  #888785  170cm      oth  1942   014356555
57   2012  NaN  2028  #623a2f  165cm      oth  1966   558908746
58   2014  127  2020  #6b5442  188cm      oth  1994     #4f5b92
59   2020  NaN  2026  #18171d  153cm      brn  1932   101990935
60   2011  NaN  2029  #c0946f  157cm      utc  1981   545992967
61   2018  NaN  2027  #602927  178cm  

### Part 2

In [90]:
pd.DataFrame(valid_passports_2)

byr   eyr      hcl   iyr        pid    hgt  ecl  cid
0    1991  2022  #341e13  2016  729933757  167cm  gry  NaN
1    1930  2022  #fffffd  2018  781914826  172cm  gry  NaN
2    1942  2028  #08df7e  2011  319110455  186cm  grn  141
3    1928  2030  #fffffd  2013  705547886  168cm  grn  NaN
4    1943  2029  #602927  2019  016251942  163cm  oth  219
5    1942  2026  #efcc98  2010  117647952  184cm  gry  NaN
6    1962  2027  #888785  2011  362697676  154cm  blu  243
7    1965  2030  #435634  2016  779104554  154cm  blu  NaN
8    1988  2030  #cfa07d  2016  951967790  167cm  grn  NaN
9    1941  2028  #341e13  2016  806979833  179cm  brn   83
10   2000  2025  #866857  2018  203320330   70in  oth  NaN
11   1988  2020  #6b5442  2017  173148327  173cm  hzl  330
12   1972  2027  #efcc98  2015  838964596  154cm  oth  163
13   1976  2028  #623a2f  2020  853714682  181cm  blu  NaN
14   1957  2025  #a97842  2013  511588647  173cm  blu  225
15   1993  2028  #efcc98  2016  871652492  177cm  oth  220
16   1973  2028  #733820  2018  018982018  165cm  hzl   57
17   1960  2028  #9d2ec4  2011  149288934   63in  blu  NaN
18   1923  2024  #7d3b0c  2014  705818464  193cm  oth  221
19   1967  2030  #ceb3a1  2011  117111015  157cm  blu  NaN
20   1980  2020  #f5dfb9  2018  787186482  188cm  brn  NaN
21   1945  2023  #07c809  2013  250679100  169cm  gry  168
22   1965  2023  #6b5442  2013  683134187  171cm  hzl  NaN
23   1952  2028  #602927  2016  720794393  180cm  blu  314
24   1982  2027  #ddbf30  2016  185583837  156cm  brn  NaN
25   1988  2026  #ceb3a1  2013  987624973  175cm  grn  NaN
26   1974  2028  #a97842  2015  350988773  160cm  oth  NaN
27   1970  2028  #b6652a  2013  717504683  156cm  gry  NaN
28   1986  2025  #a97842  2015  453874703  175cm  hzl  132
29   1959  2026  #7d3b0c  2015  580659686  154cm  brn  NaN
30   1994  2023  #c0946f  2018  681417707  191cm  amb  NaN
31   1987  2022  #a97842  2018  302326561  184cm  blu   89
32   1959  2020  #866857  2011  457081226  159cm  blu  NaN
33   1964  2026  #c0946f  2012  524745721  176cm  blu  NaN
34   1974  2027  #efcc98  2014  679155617  189cm  blu   88
35   1935  2029  #733820  2020  509009432  190cm  blu  NaN
36   1931  2024  #fffffd  2019  446558583  159cm  brn  146
37   1999  2023  #0441c6  2017  068138358  175cm  blu  197
38   1992  2022  #866857  2018  905992465  193cm  brn  NaN
39   1985  2030  #c0946f  2010  616639221  161cm  gry  NaN
40   1937  2023  #a97842  2011  594747312  172cm  hzl  260
41   1928  2028  #fffffd  2015  134536806  157cm  brn   83
42   1934  2027  #583d02  2016  035433923  180cm  amb  195
43   1983  2020  #733820  2018  087167304  174cm  hzl  NaN
44   1972  2020  #888785  2013  593301831  188cm  hzl  NaN
45   1942  2029  #888785  2020  014356555  170cm  oth  282
46   1966  2028  #623a2f  2012  558908746  165cm  oth  NaN
47   1932  2026  #18171d  2020  101990935  153cm  brn  NaN
48   1929  2027  #602927  2018  256056759  178cm  grn  NaN
49   1971  2027  #20799c  2016  813526512  185cm  blu  NaN
50   1979  2026  #53b66c  2013  871628606  179cm  grn  269
51   1972  2023  #af82df  2019  908462345  183cm  gry   51
52   1951  2027  #c0946f  2011  191542220  190cm  oth  NaN
53   1988  2028  #a97842  2018  481250123  165cm  hzl  328
54   1981  2021  #602927  2011  902964474  154cm  gry  156
55   1997  2023  #623a2f  2016  339892714  156cm  brn  315
56   1960  2024  #888785  2014  561655785   73in  brn  184
57   1973  2028  #b6652a  2020  406853460  179cm  brn  346
58   1968  2026  #ceb3a1  2013  001798001  157cm  gry  235
59   1990  2021  #b6652a  2018  504383643  151cm  brn  NaN
60   1998  2026  #888785  2015  045813631  164cm  grn  237
61   1979  2030  #733820  2015  507769390  172cm  gry  212
62   1984  2024  #efcc98  2015  968062470  176cm  brn  138
63   1941  2028  #733820  2015  431922630  160cm  oth  NaN
64   1969  2023  #602927  2017  720065302   67in  grn  294
65   2002  2024  #7d3b0c  2012  296913847  174cm  gry  NaN
66   1958  2021  #602927  2012  064919306  170cm  brn